src/snowflake/snowpark/_internal/compiler/utils.py

```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+from typing import Dict, List
+
+from snowflake.snowpark._internal.analyzer.expression import Attribute
+from snowflake.snowpark._internal.analyzer.snowflake_plan import SnowflakePlan
+from snowflake.snowpark._internal.analyzer.snowflake_plan_node import (
+    SnowflakeCreateTable,
+)
+from snowflake.snowpark._internal.compiler.query_generator import QueryGenerator
+
+
+def create_query_generator(plan: SnowflakePlan) -> QueryGenerator:
+    """
+    Helper function to construct the query generator for a given valid SnowflakePlan.
+    """
+    table_create_child_attribute_map: Dict[str, List[Attribute]] = {}
+    if plan.source_plan is not None and isinstance(
+        plan.source_plan, SnowflakeCreateTable
+    ):
```

Comment:
df.collect(), df.show(), df.count() will not have a `SnowflakeCreateTable` node that the top. I believe it only happens when we have df.write... or df.cache result where this assumption will hold

Reply:
reading your tests, it seems your assumption here is only for the case when table is being written and you need attributes. then I think this assumption is fair

Reply:
the  table_create_child_attribute_map is only needed to handle SnowflakeCreateTable, here we are extracting the table_create_child_attribute_map if root node is SnowflakeCreateTable, and our assumption is that the SnowflakeCreateTable must be a root node of a logical plan, which is true today. 
Added more comment on top of the if